core/Pos/Core/Genesis/Canonical.hs

```diff
@@ -0,0 +1,435 @@
+            ]
+
+instance Monad m => ToObjectKey m StakeholderId where
+    toObjectKey = pure . formatToString hashHexF
+
+instance Monad m => ToJSON m Coin where
+    toJSON = toJSON @_ @Word64 . unsafeGetCoin  -- i. e. String
+
+instance Monad m => ToJSON m CoinPortion where
+    toJSON = toJSON @_ @Word64 . getCoinPortion  -- i. e. String
+
+-- JSNum is ok here, current timestamp is almost 6x less than maximal
+-- bound, so overflow will happen after we die.
+instance Monad m => ToJSON m Timestamp where
+    toJSON = pure . JSNum . fromIntegral
+
+instance Monad m => ToObjectKey m Address where
+    toObjectKey = pure . formatToString addressF
+
+instance Monad m => ToJSON m Address where
+    toJSON = fmap JSString . toObjectKey
+
+instance Monad m => ToJSON m (ProxySecretKey EpochIndex) where
+    toJSON ProxySecretKey {..} =
+        -- omega is encoded as a number, because in genesis we always
+        -- set it to 0.
+        mkObject
+            [ ("omega", pure (JSNum . fromIntegral $ pskOmega))
+            , ("issuerPk", toJSON pskIssuerPk)
+            , ("delegatePk", toJSON pskDelegatePk)
+            , ("cert", toJSON pskCert)
+            ]
+
+instance Monad m => ToJSON m SoftforkRule where
+    toJSON SoftforkRule {..} =
+        mkObject
+            [ ("initThd", toJSON srInitThd)
+            , ("minThd", toJSON srMinThd)
+            , ("thdDecrement", toJSON srThdDecrement)
+            ]
+
+instance Monad m => ToJSON m Coeff where
+    toJSON (Coeff (MkFixed integer)) = toJSON @_ @Integer integer
+
+instance Monad m => ToJSON m TxFeePolicy where
+    toJSON (TxFeePolicyTxSizeLinear (TxSizeLinear summand multiplier)) =
+        mkObject
+            [("summand", toJSON summand), ("multiplier", toJSON multiplier)]
+    toJSON (TxFeePolicyUnknown {}) =
+        error "Having TxFeePolicyUnknown in genesis is likely a bug"
+
+instance Monad m => ToJSON m GenesisWStakeholders where
+    toJSON (GenesisWStakeholders stks) = toJSON stks
+
+instance Monad m => ToJSON m GenesisDelegation where
+    toJSON = toJSON . toList . unGenesisDelegation
+
+instance Monad m => ToJSON m ProtocolConstants where
+    toJSON ProtocolConstants {..} =
+        mkObject
+            -- 'k' definitely won't exceed the limit
+            [ ("k", pure . JSNum . fromIntegral $ pcK)
+            , ("protocolMagic", toJSON pcProtocolMagic)
+            , ("vssMaxTTL", toJSON pcVssMaxTTL)
+            , ("vssMinTTL", toJSON pcVssMinTTL)
+            ]
+
+instance Monad m => ToJSON m GenesisAvvmBalances where
+    toJSON = toJSON . getGenesisAvvmBalances
+
+instance Monad m => ToJSON m SharedSeed where
+    toJSON (SharedSeed seed) = pure . JSString . formatToString base16F $ seed
+
+instance Monad m => ToJSON m BlockVersionData where
+    toJSON (BlockVersionData scriptVersion slotDuration maxBlockSize maxHeaderSize maxTxSize maxProposalSize mpcThd heavyDelThd updateVoteThd updateProposalThd updateImplicit softforkRule txFeePolicy unlockStakeEpoch) =
```

Comment:
Lol, Vanya wrote it not to forget field when it would be added: compiler will say you forgot field in constructor and you will be forced to fix this instance. if you use wildcard you probably will forget to fix this instance

Reply:
Yes, exactly. I actually could have used `RecordWildCards` here, because we also have tests for this stuff, but it's safer to have extra protection apart from tests.